pkg/tmpl/context_tmpl.go

```diff
@@ -2,51 +2,140 @@ package tmpl
 	aliased := template.FuncMap{}
 
+	// map of function aliases
 	aliases := map[string]string{
 		"get": "sprigGet",
 	}
 
+	// get the default sprig functions
 	funcMap := sprig.TxtFuncMap()
 
+	// add aliases to the aliased FuncMap
 	for orig, alias := range aliases {
 		aliased[alias] = funcMap[orig]
 	}
 
+	// add functions from the Context's createFuncMap() method to the funcMap
 	for name, f := range c.createFuncMap() {
 		funcMap[name] = f
 	}
 
+	// add aliased functions to the funcMap
 	for name, f := range aliased {
 		funcMap[name] = f
 	}
 
 	return funcMap
 }
 
-func (c *Context) newTemplate() *template.Template {
+type tplInfo struct {
+	name    string
+	content string
+}
+
+// helperTPLs returns the contents of all files with names starting with "_" and ending with ".tpl"
+// in the root directory of the Context. It reads each file and appends its content to the contents slice.
+// If any error occurs during the file reading or globbing process, it returns an error.
+func (c *Context) helperTPLs() ([]tplInfo, error) {
+	tplInfos := []tplInfo{}
+	wd, err := c.fs.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get working directory: %v", err)
+	}
+	files, err := c.fs.Glob(filepath.Join(wd, "_*.tpl"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to glob helper templates: %v", err)
+	}
+	for _, file := range files {
+		content, err := c.fs.ReadFile(file)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read helper template %s: %v", file, err)
+		}
+		tplInfos = append(tplInfos, tplInfo{name: file, content: string(content)})
+	}
+	return tplInfos, nil
+}
+
+// newTemplate creates a new template based on the context.
+// It initializes the template with the specified options and parses the helper templates.
+// It also adds the 'include' function to the template's function map.
+// The 'include' function allows including and rendering nested templates.
+// The function returns the created template or an error if any occurred.
+func (c *Context) newTemplate() (*template.Template, error) {
 	funcMap := c.CreateFuncMap()
 
-	tmpl := template.New("stringTemplate").Funcs(funcMap)
+	tmpl := template.New("stringTemplate")
 	if c.preRender {
 		tmpl = tmpl.Option("missingkey=zero")
 	} else {
 		tmpl = tmpl.Option("missingkey=error")
 	}
-	return tmpl
+
+	tpls, err := c.helperTPLs()
+	if err != nil {
+		return nil, err
+	}
+	for _, tpl := range tpls {
+		tmpl, err = tmpl.Parse(tpl.content)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse helper template %s: %v", tpl.name, err)
+		}
+	}
+
+	includedNames := make(map[string]int)
+
+	// Add the 'include' function here so we can close over t.
+	funcMap["include"] = func(name string, data interface{}) (string, error) {
+		var buf strings.Builder
+		if v, ok := includedNames[name]; ok {
+			if v > recursionMaxNums {
+				return "", errors.Wrapf(fmt.Errorf("unable to execute template"), "rendering template has a nested reference name: %s", name)
```

Comment:
Good job!
A test case for this error would make it perfect!

Reply:
@mumoshu that logic comes from helm. I just copied and changed a bit.

Reply:
It doesn't justify not having a test! We might change or refactor it in the future and test prevents breakage.

Reply:
@mumoshu make sense.